gettext-maps/build.rs

```diff
@@ -0,0 +1,149 @@
+                .arg(format!("--output-file={}", mo_file_path.to_str().unwrap()))
+                .arg(&po_file_path)
+                .output()
+        };
+
+        if let Ok(metadata) = std::fs::metadata(&mo_file_path) {
+            // MO file exists, but might be outdated.
+            let mo_mtime = metadata.modified().unwrap();
+            let po_mtime = dir_entry.metadata().unwrap().modified().unwrap();
+            if mo_mtime < po_mtime {
+                update_mo_file().unwrap();
+            };
+        } else {
+            // MO file does not exist.
+            update_mo_file().unwrap();
+        }
+
+        // At this point we have an up-to-date MO file.
+        // We proceed to parse the file into a map, which is put into a Rust source file,
+        // such that it can be used from other source code.
+
+        let mut mo_file = File::open(&mo_file_path).unwrap();
+        let mut file_content = vec![];
+        mo_file.read_to_end(&mut file_content).unwrap();
+        let language_localizations = parse_mo_file(&file_content).unwrap();
+
+        let mut single_language_localization_map = phf_codegen::Map::new();
+
+        // The values will be written into the source code as is, meaning escape sequences and
+        // double quotes in the data will be interpreted by the Rust compiler, which is undesirable.
+        // Converting them to raw strings prevents this. (As long as no input data contains `"###`.)
+        fn to_raw_str(s: &str) -> String {
+            format!("r###\"{s}\"###")
+        }
+        for (k, v) in language_localizations {
+            single_language_localization_map.entry(
+                String::from_utf8(k.into()).unwrap(),
+                to_raw_str(&String::from_utf8(v.into()).unwrap()),
+            );
+        }
+        // TODO(MSRV >= 1.74): remove this. Workspace lints allow non_upper_case_globals, but
+        // workspace lints are not supported prior to Rust 1.74.
+        // https://doc.rust-lang.org/cargo/reference/workspaces.html#the-lints-table
+        writeln!(
+            &mut localization_map_file,
+            "#[allow(non_upper_case_globals)]",
+        )
+        .unwrap();
+        // Each language gets its own static map for the mapping from message in the source code to
+        // the localized version.
+        let map_name = format!("LANG_MAP_{language}");
+        write!(
+            &mut localization_map_file,
+            "static {}: phf::Map<&'static str, &'static str> = {}",
+            &map_name,
+            single_language_localization_map.build()
+        )
+        .unwrap();
+        writeln!(&mut localization_map_file, ";").unwrap();
+
+        // Map from the language identifier to the map containing the localizations for this
+        // language.
+        catalogs.entry(language, format!("&{map_name}"));
+    }
+
+    write!(
+        &mut localization_map_file,
+        "pub static CATALOGS: phf::Map<&str, &phf::Map<&str, &str>> = {}",
```

Comment:
so neither this nor the `rust-embed` approach in the other PR are currently able to compress the strings, which should explain the ~5MB binary size overhead.
I guess since existing "*.mo" files are compressed, it seems like a good idea to keep doing that, moreso because most users will use at most one locale.
If we compress, it becomes harder to use a compile-time map.. but a relatively simple approach is to store compressed data, uncompress it at runtime and build the map then.
In theory, something like rust-embed should be able to use compression.

So I'm not sure which approach is the best, I haven't reviewed this PR in detail